pep-0670.rst

```diff
@@ -196,74 +193,81 @@ Using static inline functions in the internal C API is fine: the
 internal C API exposes implementation details by design and should not be
 used outside Python.
 
-Cast to PyObject*
------------------
+Cast pointer arguments
+----------------------
+
+Existing cast
+^^^^^^^^^^^^^
+
+Currently, most macros accepting pointers cast pointers arguments to the
+expected types. For example, in Python 3.6, the ``Py_TYPE()`` macro
+casts its argument to ``PyObject*``::
 
-When a macro is converted to a function and the macro casts its
-arguments to ``PyObject*``, the new function comes with a new macro
-which cast arguments to ``PyObject*`` to prevent emitting new compiler
-warnings.  This implies that a converted function will accept pointers to
-structures inheriting from ``PyObject`` (ex: ``PyTupleObject``).
+    #define Py_TYPE(ob) (((PyObject*)(ob))->ob_type)
 
-For example, the ``Py_TYPE(obj)`` macro casts its ``obj`` argument to
-``PyObject*``::
+The ``Py_TYPE()`` macro accept ``PyObject*`` type, but also any pointer
+type like ``PyLongObject*`` or ``PyDictObject*``.
 
-    #define _PyObject_CAST_CONST(op) ((const PyObject*)(op))
+Add a new macro to keep the cast
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-    static inline PyTypeObject* _Py_TYPE(const PyObject *ob) {
+When a macro is converted to a function and the macro casts at least one
+of its argument, a new macro is added to keep the cast. The new macro
+and the function have the same name. Example with the ``Py_TYPE()``
+macro converted to a static inline function::
+
+    static inline PyTypeObject* Py_TYPE(PyObject *ob) {
         return ob->ob_type;
     }
-    #define Py_TYPE(ob) _Py_TYPE(_PyObject_CAST_CONST(ob))
+    #define Py_TYPE(ob) Py_TYPE((PyObject*)(ob))
 
-The undocumented private ``_Py_TYPE()`` function must not be called
-directly. Only the documented public ``Py_TYPE()`` macro must be used.
+The cast is kept all pointer types, not only ``PyObject*``.
```

Comment:
```suggestion
The cast is kept for all pointer types, not only ``PyObject*``.
```

Grammar error causing semantic ambiguity; I _believe_ this is what you intended to say.